cpp/src/Ice/RouterInfo.cpp

```diff
@@ -144,45 +144,29 @@ IceInternal::RouterInfo::getClientEndpoints()
 }
 
 void
-IceInternal::RouterInfo::getClientProxyResponse(const Ice::ObjectPrxPtr& proxy,
-                                                const optional<bool>& hasRoutingTable,
-                                                const GetClientEndpointsCallbackPtr& callback)
-{
-    callback->setEndpoints(setClientEndpoints(proxy, hasRoutingTable ? hasRoutingTable.value() : true));
-}
-
-void
-IceInternal::RouterInfo::getClientProxyException(std::exception_ptr ex,
-                                                 const GetClientEndpointsCallbackPtr& callback)
-{
-    callback->setException(ex);
-}
-
-void
-IceInternal::RouterInfo::getClientEndpoints(const GetClientEndpointsCallbackPtr& callback)
+IceInternal::RouterInfo::getClientEndpointsAsync(
+    std::function<void(vector<EndpointIPtr>)> response,
+    std::function<void(std::exception_ptr)> ex)
 {
     vector<EndpointIPtr> clientEndpoints;
     {
         lock_guard lock(_mutex);
         clientEndpoints = _clientEndpoints;
     }
 
-    if(!clientEndpoints.empty())
+    if (!clientEndpoints.empty())
     {
-        callback->setEndpoints(clientEndpoints);
+        response(std::move(clientEndpoints));
         return;
     }
 
     RouterInfoPtr self = shared_from_this();
     _router->getClientProxyAsync(
-        [self, callback](const Ice::ObjectPrxPtr& proxy, optional<bool> hasRoutingTable)
+        [self, response](const Ice::ObjectPrxPtr& proxy, optional<bool> hasRoutingTable)
         {
-            self->getClientProxyResponse(proxy, hasRoutingTable, callback);
+            response(self->setClientEndpoints(proxy, hasRoutingTable ? hasRoutingTable.value() : true));
```

Comment:
Can we use `value_or` here?

```suggestion
            response(self->setClientEndpoints(proxy, hasRoutingTable.value_or(true)));
```